ide/nativeimage.api/src/org/netbeans/modules/nativeimage/api/debug/NIDebugger.java

```diff
@@ -113,6 +113,20 @@ public void removeBreakpoint(Object id) {
         return provider.start(command, workingDirectory, debugger, displayName, executionDescriptor, startedEngine);
     }
 
+    /**
+     * Attach to a process and create a debugging session. Call this typically after breakpoints are added.
+     *
+     * @param executablePath path to an executable representing the native image
+     * @param processId a process to attach to
+     * @param debugger the native debugger command
+     * @param startedEngine the corresponding DebuggerEngine is passed to this consumer
+     * @return future that completes on the execution finish
+     * @since 0.4
+     */
+    public CompletableFuture<Void> attach(String executablePath, long processId, String debugger, Consumer<DebuggerEngine> startedEngine) {
```

Comment:
Wouldn't it be better to return `CompletionStage<DebuggerEngine>` and remove the 4th parameter? Probably not, as the `Consumer<DebuggerEngine>` is called way sooner than the `CompletableFuture<Void>` is set, rigth?

Reply:
Right. The consumer may be called even before the `attach` method returns, depending on the actual threads scheduling.